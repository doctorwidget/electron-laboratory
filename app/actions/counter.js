// @flow
/**
 * Switching to redux-saga has the happy benefit of eliminating this weird
 * dependency on things like {type, Dispatch, GetState} from the flow library
 * import type { GetState, Dispatch } from '../reducers/types';
 */

export const INCREMENT_COUNTER = 'INCREMENT_COUNTER';
export const DECREMENT_COUNTER = 'DECREMENT_COUNTER';

export const INCREMENT_IF_EVEN = 'INCREMENT_IF_EVEN';
export const INCREMENT_IF_ODD = 'INCREMENT_IF_ODD';

export const INCREMENT_ASYNC = 'INCREMENT_ASYNC';
export const DECREMENT_ASYNC = 'DECREMENT_ASYNC';

export function increment() {
  return {
    type: INCREMENT_COUNTER
  };
}

export function decrement() {
  return {
    type: DECREMENT_COUNTER
  };
}

/** Original thunk version of incrementIfOdd
 * How would you test this in Jest?
 * This function returns another function.
 * That function needs both dispatch and getState() to do anything!
 * When even, _nothing happens_
 * When odd, another action gets generated.
 * The whole thing is a poor candidate for unit testing.
export function incrementIfOdd() {
  return (dispatch: Dispatch, getState: GetState) => {
    const { counter } = getState();

    if (counter % 2 === 0) {
      return;
    }

    dispatch(increment());
  };
}
*/

/** Original thunk version of incrementAsync -- now not used at all 
 * Hard to test: see {project}/test/actions/counter.js for the commented-out test code
export function incrementAsync(delay: number = 1000) {
  return (dispatch: Dispatch) => {
    setTimeout(() => {
      dispatch(increment());
    }, delay);
  };
}
*/

/**
 * Redux-saga-centric actions follow -- not part of the original boilerplate.
 * In this pattern, the logic stays _out_ of the action creators, and moves
 * to the sagas.js files. Contrast that with incrementIfOdd && incrementAsync
 * above, which are thunk-centric, putting logic into the action creators file.
 *
 * Every Action type must have an associated action creator if we want it to
 * participate in the bindActionCreators shortcut. However, I am still on the
 * fence about whether or not that shortcut is worth the loss of clarity it
 * brings along with it.
 */

/**
 * This action type is intended to be heard by redux-saga
 */
export function incrementIfEven() {
  return {
    type: INCREMENT_IF_EVEN
  };
}

/**
 * Redux-saga replacement for original thunk incrementIfOdd
 *
 * _Requests_ that we increment... maybe!
 * Logic is not handled within the action creator: this is a big advantage
 * redux-saga has over redux-thunk, IMHO. Action creators stay puny, and
 * logic is going to live in sagas, which are always testable due to
 * returning highly-predictable Effect objects.
 */
export function incrementIfOdd() {
  return {
    type: INCREMENT_IF_ODD
  };
}

/**
 * This action type is intended to be heard by redux-saga
 */
export function decrementAsync() {
  return {
    type: DECREMENT_ASYNC
  };
}

/**
 * Request for a delayed and debounced increment.
 */
export function incrementAsync() {
  return {
    type: INCREMENT_ASYNC
  };
}
