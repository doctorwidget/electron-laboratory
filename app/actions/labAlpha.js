/**
 * Action constant strings and action creator functions for the alpha page
 */

export const ALPHA_ACTIVATE = 'ALPHA_ACTIVATE';
export const ALPHA_DEACTIVATE = 'ALPHA_DEACTIVATE';

export function alphaActivate() {
  return {
    type: ALPHA_ACTIVATE
  };
}

export function alphaDeactivate() {
  return {
    type: ALPHA_DEACTIVATE
  };
}
