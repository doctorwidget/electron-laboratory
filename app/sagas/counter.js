/**
 * SAGAS for the counter.
 */
import { delay } from 'redux-saga';
import { all, put, select, takeEvery, takeLatest } from 'redux-saga/effects';

// ../actions/counter has no default action, so we must manually group it
// that's the only good reason to use 'import *': to group a mess of individual
// imports into one object rather than spraying them all into the current environment
import * as counterActions from '../actions/counter';

/**
 * Sagas tend to travel in pairs. This is a fundamental detail of the pattern!
 * Think in terms of two kinds of sagas:
 *
 * - _worker_ sagas,  which _do_ things (make async calls, navigate within the app, etc)
 * - _watcher_ sagas, which subscribe to the redux store via middleware
 *
 * The _watchers_ instantiate _workers_ in response to actions.
 */

/**
 * This __worker__ saga listens to the same INCREMENT_COUNTER action already
 * handled via thunks. At first, it will output a console message for us to
 * look at in the console. Eventually, we'll have the saga system take over
 * from the thunk system.
 */
export const doIncrement = function* doIncrement() {
  yield delay(1); // just to show we can do this without affecting anything
  console.log(`doIncrement WORKER saga logging an increment. Done.`);
  // That's a silly thing to waste code doing (for several reasons)
  // But I do want to show that sagas can listen to ALL actions
  // In other words, sagas seamlessly tap into the redux dispatcher,
  // That means sagas can be used for _all_ side effects, not just asynchronicity!
};

/**
 * The _watcher_ saga for increment events.
 * Note that this isn't actually an asynchronous operation, so there is no point
 * in handling it via a saga! I'm only throwing in this in here as a demonstration
 * that any number of listeners can be reacting to one action.
 */
export const watchIncrement = function* watchIncrement() {
  // IFF this saga is connected to the redux store via react-saga middleware
  // THENN this saga listens for INCREMENT_COUNTER actions
  // and instantiates a trialStartSaga instance in response
  yield takeEvery(counterActions.INCREMENT_COUNTER, doIncrement);
};

/** Another __worker__ saga */
export const doDecrement = function* doDecrement() {
  console.log(`doDecrement WORKER saga logging a decrement. Done.`);
  yield delay(1); // again, a useless but harmless delay
};

/**
 * The _watcher_ saga for decrement events.
 * Again, we just listen in on a pre-existing event type.
 */
export const watchDecrement = function* watchDecrement() {
  // IFF this saga is connected to the redux store via react-saga middleware
  // THENN this saga listens for INCREMENT_COUNTER actions
  // and instantiates a trialStartSaga instance in response
  yield takeEvery(counterActions.DECREMENT_COUNTER, doDecrement);
};

/**
 * Note that the boilerplate components don't have dedicated event types for
 * the special events: incrementIfOdd or incrementAsync. Both of those are
 * handled via the thunk system without creating a new event type. The logic
 * for this is inlined into the `{project}/actions/counter.js` file, which
 * seems like an antipattern to me, because it's clearly adding responsibility
 * to the action creator file beyond its original purpose.
 */

/**
 * Here we inmplement incrementIfEven as a saga. That means we create an action
 * type INCREMENT_IF_EVEN, which no _reducer_ will react to. This _saga_ is
 * the only thing currently set up to will react to that event. If it decides
 * the time is right, it will emit its own _INCREMENT_ action via `put(increment())`.
 * Now the business logic of odd/even lives in the saga, rather than the action
 * creator.
 *
 * But note that I think this pattern is problematic, and still needs some
 * fundamental thinking over, because I don't want to drift too far away from
 * the pattern where the REDUCERS have 100% of the business logic.
 *
 * Also note that this relies on using the select() Effect from redux-saga,
 * which exists explicitly for the purpose of peeking in to the current app
 * state. The official docs include warning along the lines of what I say
 * above: use this power responsibly!
 * https://redux-saga.js.org/docs/api/index.html#selectselector-args
 */
export const doIncrementIfEven = function* doDecrementIfEven() {
  // The {select} effect is the anointed way for redux-saga to peek at the current state
  // We could pass in an actual `reselect` selector function as the first argument.
  // By providing no arguments at all, we get the full current state back.
  const currentState = yield select(); // nb: the {yield} is required here!
  if (currentState.counter % 2 === 0) {
    console.log(
      'doIncrementIfEven WORKER saga: counter is even. Incrementing!'
    );
    yield put(counterActions.increment()); // nb: we put an instantiated action object, not an action creator function
  } else {
    console.log('doIncrementIfEven WORKER counter is not even. Doing nothing!');
  }
};

export const watchIncrementIfEven = function* watchIncrementIfEven() {
  yield takeEvery(counterActions.INCREMENT_IF_EVEN, doIncrementIfEven);
};

/*
 * Similarly, implement a decrementRequest saga, triggered by a new action type
 * named DECREMENT_REQUEST. Again, a saga in this file will hear that
 * request, cause the delay, and then emit the same decrement() action that
 * already exists, via `put(decrement())`. Again, business logic moves from
 * the action creators file to the sagas file.
 * 
 * I like this name better than the boilerplate example of "incrementAsync".
 * Here we emphasize that a _request_ has gone out, and _someone_ or _something_
 * has to decide whether or not to approve of it. That's an utterly classic
 * side effect concern. It might get resolved purely locally, without any kind
 * of heavyweight async operation like a network call. But it's still worth
 * pulling out as a side effect, rather than having it hard-coded into one
 * reducer somewhere. Keep the reducers focused on their own pure state logic,
 * and let messy real-world concerns live as side effects in sagas.
 */
export const doDecrementRequest = function* doDecrementRequest() {
  // in this case we'll just wait a second and then approve the request
  console.log('doDecrementRequest: considering the request');
  yield delay(2000);
  console.log('doDecrementRequest: Your request has been approved!');
  yield put(counterActions.decrement());
};

/**
 * Then as a final demonstration twist, we'll use {takeLatest} instead
 * of {takeEvery}, effectively _debouncing_ this action. If the user spams
 * the decrementAsync button over and over, nothing will happen until
 * _after_ they cease their incessant clicking. Then and only then, the
 * last known decrementRequest action will be allowed to run to completion.
 * Don't reward rapid clickers!
 */
export const watchDecrementRequest = function* watchDecrementRequest() {
  yield takeLatest(counterActions.DECREMENT_REQUEST, doDecrementRequest);
};

/**
 * Bundle up _all_ trial-related sagas into one.
 * It's never enough to declare a worker-watcher dyad above!
 * You must always also include the watcher in the final exported saga!
 */
const allCounterSagas = function* allCounterSagas() {
  // the {all} helper takes an array of saga _instances_. Not an array of saga functions!
  // It will run them all in parallel. If any of them are infinite, this saga will also
  // be infinite. This saga will only terminate if/when _all_ of the sagas have terminated.
  yield all([
    watchIncrement(), // nb: array is full of actual generator instances, not saga functions
    watchDecrement(),
    watchIncrementIfEven(),
    watchDecrementRequest()
    // add more sagas here as necessary
  ]);

  // The {all} effect is a convenience wrapper added after early versions of react-saga.
  // Before that, you would yield a simple array of sagas, each wrapped in a {fork}, e.g.:
  //
  //     yield [
  //          fork(watchDecrementIfOdd),  // nb: the saga function, not a generator instance
  //          fork(watchDecrementIfEven)
  //     ]
  //
  // Notice there we pass in the saga _functions_, not saga instances.
  // So think of {all} as relatively late syntactic sugar for this {fork} pattern
};
export default allCounterSagas; // and make it the default export

// nb: the _user_ of this file doesn't have to use the exact phrase 'allCounterSaga'
// see ./index.js, where we just refer to this as "counterSaga"
// (which is a better name, but I wanted to demonstrate this naming flexibility)
