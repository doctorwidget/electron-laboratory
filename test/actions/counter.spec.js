/**
 * Original tests had to deal with thunks, and used sinon for this
 * I think we have now removed all need for sinon... 
 * (and doesn't Jest handle that out of the box anyway?)
  import { spy } from 'sinon';
 */
import * as actions from '../../app/actions/counter';

describe('actions', () => {
  it('should increment should create increment action', () => {
    // nb: the boilerplate uses toMatchSnapshot on the return object from increment()
    // it turns out this works! jest.toMatchSnapshot can treat an object hash just like a JSX output
    // However this seems too clever by half
    // I mean, what are you really testing? That there's been no change increment() output...
    // but wouldn't it be better to say something more explicit?
    expect(actions.increment()).toMatchSnapshot();
    // something like this!
    expect(actions.increment().type).toEqual(actions.INCREMENT_COUNTER);
  });

  it('should decrement should create decrement action', () => {
    // the boilerplate clever test:
    expect(actions.decrement()).toMatchSnapshot();
    // our own less-code-golfish test:
    expect(actions.decrement().type).toEqual(actions.DECREMENT_COUNTER);
  });

  /** Thunks replaced by sagas tests 
  it('should incrementIfOdd should create increment action', () => {
    const fn = actions.incrementIfOdd();
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    const getState = () => ({ counter: 1 });
    fn(dispatch, getState);
    expect(dispatch.calledWith({ type: actions.INCREMENT_COUNTER })).toBe(true);
  });
  */

  /** Thunks replaced by sagas tests
  it('should incrementIfOdd shouldnt create increment action if counter is even', () => {
    const fn = actions.incrementIfOdd();
    const dispatch = spy();
    const getState = () => ({ counter: 2 });
    fn(dispatch, getState);
    expect(dispatch.called).toBe(false);
  });
  */

  it('incrementIfOdd should always create an incrementIfOdd action', () => {
    // confirmed that I can create new tests with new action creators,
    // and Jest will fail the test if I subsequently change the string constant name of the action
    // So yes, this approach is _an option_
    // And it might even be a good option once you start decorating actions with payloads
    expect(actions.incrementIfOdd()).toMatchSnapshot();
    // But I kind of feel like the whole point of unit tests is to accept some WET code
    // in exchange for dirt-simple reassurance that your code is doing exactly what you think it is
    // In other words, not "does this output look like it did whenever I last committed snapshots"
    // but instead "does this output look exactly like what I type next"
    expect(actions.incrementIfOdd().type).toEqual(actions.INCREMENT_IF_ODD);
  });

  /** 
  * Original test for incrementAsync
  * The action test is now clean and simple, focused only on the action itself.
  * See the ../sagas/counter.spec.js file for saga logic testing
  // There's no nice way to test this at the moment...
  it('should incrementAsync', done => {
    const fn = actions.incrementAsync(1);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    setTimeout(() => {
      expect(dispatch.calledWith({ type: actions.INCREMENT_COUNTER })).toBe(
        true
      );
      done();
    }, 5);
  });
  */

  it('incrementAsync should always create an incrementAsync action', () => {
    expect(actions.incrementAsync().type).toEqual(actions.INCREMENT_ASYNC);
  });
});
