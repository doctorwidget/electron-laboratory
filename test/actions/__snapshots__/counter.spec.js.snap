// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions incrementIfOdd should always create an incrementIfOdd action 1`] = `
Object {
  "type": "INCREMENT_IF_ODD",
}
`;

exports[`actions should decrement should create decrement action 1`] = `
Object {
  "type": "DECREMENT_COUNTER",
}
`;

exports[`actions should increment should create increment action 1`] = `
Object {
  "type": "INCREMENT_COUNTER",
}
`;
